// enka-network-api/dist/models/artifact/ArtifactSetBonus.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
const StatProperty_1 = __importDefault(require("../StatProperty"));
/**
 * @en ArtifactSetBonus
 */
class ArtifactSetBonus {
    /**
     * @param needCount
     * @param data
     * @param enka
     */
    constructor(needCount, data, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("affixId");
        this.needCount = needCount;
        this.description = new TextAssets_1.default(json.getAsNumber("descTextMapHash"), enka);
        this.addProps = json.get("addProps").filterArray((_, p) => p.has("propType") && p.has("value")).map(([, p]) => new StatProperty_1.default(p.getAsString("propType"), p.getAsNumber("value"), enka));
        this.paramList = json.get("paramList").mapArray((_, p) => p.getAsNumber());
    }
}
exports.default = ArtifactSetBonus;
