// enka-network-api/dist/models/assets/ImageAssets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const imageBaseUrlMihoyo = "https://upload-os-bbs.mihoyo.com/game_record/genshin";
const imageTypes = {
    "": [/^UI_AvatarIcon_(.+)_Circle$/],
    "character_side_icon": [/^UI_AvatarIcon_Side_(.+)$/],
    "character_icon": [/^UI_AvatarIcon_(.+)$/],
    "equip": [/^UI_EquipIcon_(.+?)(_Awaken)?$/, /^UI_RelicIcon_(.+)$/],
    // "constellation_icon": [/^UI_Talent_(.+)$/],
};
/**
 * @en ImageAssets
 */
class ImageAssets {
    /**
     * @param name
     * @param enka
     */
    constructor(name, enka) {
        var _a;
        this.enka = enka;
        this.name = name;
        this.imageBaseUrl = (_a = Object.keys(enka.options.imageBaseUrlByRegex).find(url => enka.options.imageBaseUrlByRegex[url].some(regex => regex.test(name)))) !== null && _a !== void 0 ? _a : enka.options.defaultImageBaseUrl;
        this.url = name === "" ? "" : `${this.imageBaseUrl}/${name}.png`;
        this.imageType = Object.keys(imageTypes).find(type => imageTypes[type].some(regex => regex.test(name))) || null;
        this.mihoyoUrl = (name === "" || !this.imageType) ? "" : `${imageBaseUrlMihoyo}/${this.imageType}/${name}.png`;
        this.isAvailable = this.name !== null && this.name !== undefined && this.name !== "";
    }
}
exports.default = ImageAssets;
