// enka-network-api/dist/models/material/Material.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NameCard = void 0;
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
const ImageAssets_1 = __importDefault(require("../assets/ImageAssets"));
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
/**
 * @en Material
 */
class Material {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        this._data = data;
        this.enka = enka;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.name = new TextAssets_1.default(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.default(json.getAsNumber("descTextMapHash"), enka);
        this.icon = new ImageAssets_1.default(json.getAsString("icon"), enka);
        this.pictures = json.get("picPath").mapArray((_, name) => new ImageAssets_1.default(name.getAsString(), enka));
        this.itemType = json.getAsString("itemType");
        this.materialType = json.getAsStringWithDefault(null, "materialType");
        this.stars = json.getAsNumberWithDefault(null, "rankLevel");
    }
    /**
     * @param data
     * @param enka
     */
    static getMaterialByData(data, enka) {
        const json = new config_file_js_1.JsonReader(data);
        switch (json.getAsStringWithDefault(null, "materialType")) {
            case NameCard.MATERIAL_TYPE:
                return new NameCard(data, enka);
            default:
                return new Material(data, enka);
        }
    }
    /**
     * @param id
     * @param enka
     */
    static getMaterialById(id, enka) {
        var _a;
        if (isNaN(Number(id)))
            throw new Error("Parameter `id` must be a number or a string number.");
        id = Number(id);
        const materialData = (_a = enka.cachedAssetsManager.getGenshinCacheData("MaterialExcelConfigData").findArray((_, p) => p.getAsNumber("id") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!materialData)
            throw new AssetsNotFoundError_1.default("Material", id);
        return this.getMaterialByData(materialData.getAsJsonObject(), enka);
    }
}
exports.default = Material;
/**
 * @en NameCard
 * @extends {Material}
 */
class NameCard extends Material {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        super(data, enka);
        this.materialType = "MATERIAL_NAMECARD";
    }
}
exports.NameCard = NameCard;
/**  */
NameCard.MATERIAL_TYPE = "MATERIAL_NAMECARD";
