// enka-network-api/dist/models/material/UpgradeCost.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const Material_1 = __importDefault(require("./Material"));
const ts_utils_1 = require("../../utils/ts_utils");
/**
 * @en UpgradeCost
 */
class UpgradeCost {
    /**
     * @param coinCost
     * @param costItems
     * @param enka
     */
    constructor(coinCost, costItems, enka) {
        this.enka = enka;
        this.coin = coinCost;
        const itemsJson = new config_file_js_1.JsonReader(costItems);
        this.items = itemsJson.mapArray((_, cost) => {
            if (!cost.has("id"))
                return null;
            return {
                material: Material_1.default.getMaterialById(cost.getAsNumber("id"), enka),
                count: cost.getAsNumber("count"),
            };
        }).filter(ts_utils_1.nonNullable);
    }
}
exports.default = UpgradeCost;
