// enka-network-api/dist/models/character/talents/Skill.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../../errors/AssetsNotFoundError"));
const ImageAssets_1 = __importDefault(require("../../assets/ImageAssets"));
const TextAssets_1 = __importDefault(require("../../assets/TextAssets"));
const DynamicTextAssets_1 = __importDefault(require("../../assets/DynamicTextAssets"));
/**
 * Normal Attack, Elemental Skill, and Elemental Burst. Not including Passive Talents.
 * @en Skill
 */
class Skill {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.name = new TextAssets_1.default(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new DynamicTextAssets_1.default(json.getAsNumber("descTextMapHash"), {}, enka);
        this.icon = new ImageAssets_1.default(json.getAsString("skillIcon"), enka);
    }
    /**
     * @param id
     * @param enka
     */
    static getById(id, enka) {
        return new Skill(this._getJsonObjectById(id, enka), enka);
    }
    static _getJsonObjectById(id, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("AvatarSkillExcelConfigData").findArray((_, p) => p.getAsNumber("id") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("Skill", id);
        return json.getAsJsonObject();
    }
}
exports.default = Skill;
