// enka-network-api/dist/models/character/CharacterDetails.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
const ImageAssets_1 = __importDefault(require("../assets/ImageAssets"));
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
const character_utils_1 = require("../../utils/character_utils");
const CharacterVoiceData_1 = __importDefault(require("./CharacterVoiceData"));
/**
 * @en CharacterDetails
 */
class CharacterDetails {
    /**
     * @param data
     * @param isArchon
     * @param enka
     */
    constructor(data, isArchon, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("fetterId");
        this.characterId = json.getAsNumber("avatarId");
        this._nameId = (0, character_utils_1.getNameIdByCharacterId)(this.characterId, enka);
        this.birthday = (json.has("infoBirthMonth") && json.has("infoBirthDay")) ? { month: json.getAsNumber("infoBirthMonth"), day: json.getAsNumber("infoBirthDay") } : null;
        this.location = new TextAssets_1.default(json.getAsNumber("avatarNativeTextMapHash"), enka);
        this.vision = new TextAssets_1.default(json.getAsNumber("avatarVisionBeforTextMapHash"), enka);
        this.constellation = new TextAssets_1.default(isArchon ? json.getAsNumber("avatarConstellationAfterTextMapHash") : json.getAsNumber("avatarConstellationBeforTextMapHash"), enka);
        this.constellationIcon = new ImageAssets_1.default(`Eff_UI_Talent_${this._nameId}`, enka);
        this.title = new TextAssets_1.default(json.getAsNumber("avatarTitleTextMapHash"), enka);
        this.description = new TextAssets_1.default(json.getAsNumber("avatarDetailTextMapHash"), enka);
        this.cv = {
            chinese: new TextAssets_1.default(json.getAsNumber("cvChineseTextMapHash"), enka),
            japanese: new TextAssets_1.default(json.getAsNumber("cvJapaneseTextMapHash"), enka),
            english: new TextAssets_1.default(json.getAsNumber("cvEnglishTextMapHash"), enka),
            korean: new TextAssets_1.default(json.getAsNumber("cvKoreanTextMapHash"), enka),
        };
    }
    /**
     * @param lang
     */
    getCvByLanguage(lang) {
        lang !== null && lang !== void 0 ? lang : (lang = this.enka.options.defaultLanguage);
        switch (lang) {
            case "chs":
            case "cht":
                return this.cv.chinese;
            case "jp":
                return this.cv.japanese;
            case "kr":
                return this.cv.korean;
            default:
                return this.cv.english;
        }
    }
    /**
     * @returns voice data of the character. This does not contain audio data or files.
     */
    getVoices() {
        const allVoices = this.enka.cachedAssetsManager.getGenshinCacheData("FettersExcelConfigData");
        const filtered = allVoices.filterArray((_, voiceData) => voiceData.getAsNumber("avatarId") === this.characterId);
        return filtered.map(([, voice]) => new CharacterVoiceData_1.default(voice.getAsJsonObject(), this.enka));
    }
    /**
     * @param id
     * @param isArchon
     * @param enka
     */
    static getById(id, isArchon, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("FetterInfoExcelConfigData").findArray((_, f) => f.getAsNumber("fetterId") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("FetterInfo", id);
        return new CharacterDetails(json.getAsJsonObject(), isArchon, enka);
    }
    /**
     * @param id
     * @param isArchon
     * @param enka
     */
    static getByCharacterId(id, isArchon, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("FetterInfoExcelConfigData").findArray((_, f) => f.getAsNumber("avatarId") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("FetterInfo by avatarId", id);
        return new CharacterDetails(json.getAsJsonObject(), isArchon, enka);
    }
}
exports.default = CharacterDetails;
