// enka-network-api/dist/models/character/Constellation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
const ImageAssets_1 = __importDefault(require("../assets/ImageAssets"));
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
const StatProperty_1 = __importDefault(require("../StatProperty"));
/**
 * @en Constellation
 */
class Constellation {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(data);
        this.id = json.getAsNumber("talentId");
        this.name = new TextAssets_1.default(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.default(json.getAsNumber("descTextMapHash"), enka);
        this.icon = new ImageAssets_1.default(json.getAsString("icon"), enka);
        this.addProps = json.get("addProps").filterArray((_, p) => p.has("propType") && p.has("value")).map(([, p]) => new StatProperty_1.default(p.getAsString("propType"), p.getAsNumber("value"), enka));
        this.paramList = json.get("paramList").mapArray((_, p) => p.getAsNumber());
    }
    /**
     * @param id
     * @param enka
     */
    static getById(id, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("AvatarTalentExcelConfigData").findArray((_, p) => p.getAsNumber("talentId") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("Constellation", id);
        return new Constellation(json.getAsJsonObject(), enka);
    }
}
exports.default = Constellation;
