// enka-network-api/dist/models/character/Costume.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
const ImageAssets_1 = __importDefault(require("../assets/ImageAssets"));
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
const character_utils_1 = require("../../utils/character_utils");
const CharacterData_1 = __importDefault(require("./CharacterData"));
/**
 * @en Costume
 */
class Costume {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        this.enka = enka;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("skinId");
        this.name = new TextAssets_1.default(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.default(json.getAsNumber("descTextMapHash"), enka);
        this.characterId = json.getAsNumber("characterId");
        this.isDefault = json.getAsBooleanWithDefault(false, "isDefault");
        const jsonName = json.getAsString("jsonName");
        this._nameId = !this.isDefault ? jsonName.slice(jsonName.lastIndexOf("_") + 1) : (0, character_utils_1.getNameIdByCharacterId)(this.characterId, enka);
        this.icon = new ImageAssets_1.default(`UI_AvatarIcon_${this._nameId}`, enka);
        this.sideIcon = new ImageAssets_1.default(`UI_AvatarIcon_Side_${this._nameId}`, enka);
        this.splashImage = new ImageAssets_1.default(!this.isDefault ? `UI_Costume_${this._nameId}` : `UI_Gacha_AvatarImg_${this._nameId}`, enka);
        this.stars = !this.isDefault ? json.getAsNumber("quality") : null;
        this.cardIcon = new ImageAssets_1.default(this.isDefault ? "UI_AvatarIcon_Costume_Card" : `UI_AvatarIcon_${this._nameId}_Card`, enka);
    }
    /**  */
    getCharacterData() {
        return CharacterData_1.default.getById(this.characterId, this.enka);
    }
    /**
     * @param id
     * @param enka
     */
    static getById(id, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("AvatarCostumeExcelConfigData").findArray((_, p) => p.getAsNumber("skinId") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("Costume", id);
        return new Costume(json.getAsJsonObject(), enka);
    }
    /**
     * @param characterId
     * @param enka
     */
    static getDefaultCostumeByCharacterId(characterId, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("AvatarCostumeExcelConfigData").findArray((_, p) => p.getAsNumber("characterId") === characterId && p.getAsBooleanWithDefault(false, "isDefault"))) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("Default costume", characterId);
        return new Costume(json.getAsJsonObject(), enka);
    }
}
exports.default = Costume;
