// enka-network-api/dist/models/enka/GenshinCharacterBuild.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const Character_1 = __importDefault(require("../character/Character"));
const enka_system_1 = require("enka-system");
/**
 * A present character build or a saved character build in Enka.Network account
 * @en GenshinCharacterBuild
 */
class GenshinCharacterBuild extends enka_system_1.CharacterBuild {
    /**
     * @param data
     * @param enka
     * @param username
     * @param hash
     */
    constructor(data, enka, username, hash) {
        super();
        this._data = data;
        this.enka = enka;
        this.enkaUserInfo = { username: username, hash: hash };
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.name = json.getAsString("name");
        this.order = json.getAsNumber("order");
        this.isLive = json.getAsBoolean("live");
        this.isPublic = json.getAsBoolean("public");
        this.character = new Character_1.default(json.getAsJsonObject("avatar_data"), enka);
        this.imageUrl = json.getAsNullableString("image");
        this.hoyoType = json.getAsNumber("hoyo_type");
        this.url = `${this.enka.options.enkaUrl}/u/${this.enkaUserInfo.username}/${this.enkaUserInfo.hash}/${this.character.characterData.id}/${this.id}/`;
    }
}
exports.default = GenshinCharacterBuild;
