// enka-network-api/dist/models/DetailedGenshinUser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const Character_1 = __importDefault(require("./character/Character"));
const GenshinUser_1 = __importDefault(require("./GenshinUser"));
/**
 * @extends {GenshinUser}
 */
class DetailedGenshinUser extends GenshinUser_1.default {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        super(data, enka);
        const json = new config_file_js_1.JsonReader(data);
        this.showCharacterDetails = json.has("avatarInfoList");
        this.characters = json.has("avatarInfoList") ? json.get("avatarInfoList").mapArray((_, p) => new Character_1.default(p.getAsJsonObject(), enka)) : [];
    }
    toGOOD() {
        return {
            format: "GOOD",
            version: 2,
            source: "yuko1101/enka-network-api",
            characters: this.characters.map(c => c.toGOOD()),
            artifacts: this.characters.flatMap(c => c.artifacts).map(a => a.toGOOD()),
            weapons: this.characters.map(c => c.weapon).map(w => w.toGOOD()),
        };
    }
}
exports.default = DetailedGenshinUser;
