// enka-network-api/dist/models/StatProperty.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const AssetsNotFoundError_1 = __importDefault(require("../errors/AssetsNotFoundError"));
const constants_1 = require("../utils/constants");
const TextAssets_1 = __importDefault(require("./assets/TextAssets"));
/**
 * @en StatProperty
 */
class StatProperty {
    /**
     * @param fightProp
     * @param value
     * @param enka
     * @param multiplied
     */
    constructor(fightProp, value, enka) {
        var _a;
        this.fightProp = fightProp;
        this.enka = enka;
        const propData = (_a = enka.cachedAssetsManager.getGenshinCacheData("ManualTextMapConfigData").findArray((_, p) => p.getAsString("textMapId") === fightProp)) === null || _a === void 0 ? void 0 : _a[1];
        if (!propData)
            throw new AssetsNotFoundError_1.default("Fight Prop", fightProp);
        this._propData = propData.getAsJsonObject();
        this.fightPropName = new TextAssets_1.default(propData.getAsNumber("textMapContentTextMapHash"), enka);
        this.isPercent = constants_1.percent.some(p => p === fightProp);
        this.rawValue = value;
        this.value = round(value, 8);
    }
    /**  */
    get valueText() {
        const fix = this.isPercent ? 1 : 0;
        return this.getMultipliedValue().toFixed(fix).replace(/\B(?=(\d{3})+(?!\d))/g, ",") + (this.isPercent ? "%" : "");
    }
    /**
     * Multiplies [value](#value) by 100 if it is a percentage
     */
    getMultipliedValue() {
        return this.value * (this.isPercent ? 100 : 1);
    }
    /**  */
    toString() {
        var _a;
        const name = (_a = this.fightPropName.get()) !== null && _a !== void 0 ? _a : "[Unknown]";
        return `${name}(${this.fightProp}): ${this.valueText}`;
    }
    /**
     * @param fightProp
     * @param enka
     */
    static getFightPropTextAssets(fightProp, enka) {
        var _a;
        const propData = (_a = enka.cachedAssetsManager.getGenshinCacheData("ManualTextMapConfigData").findArray((_, p) => p.getAsString("textMapId") === fightProp)) === null || _a === void 0 ? void 0 : _a[1];
        return propData ? new TextAssets_1.default(propData.getAsNumber("textMapContentTextMapHash"), enka) : null;
    }
    /**
     * @param statProperties
     * @param enka
     */
    static sumStatProperties(statProperties, enka) {
        const stats = {};
        for (const prop of statProperties) {
            if (stats[prop.fightProp] === undefined) {
                stats[prop.fightProp] = 0;
            }
            stats[prop.fightProp] += prop.value;
        }
        return Object.entries(stats).map(([fightProp, value]) => new StatProperty(fightProp, value, enka));
    }
}
exports.default = StatProperty;
function round(x, decimalPlaces = 0) {
    const p = Math.pow(10, decimalPlaces);
    return Math.round(x * p) / p;
}
