// enka-network-api/dist/models/weapon/WeaponRefinement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
const StatProperty_1 = __importDefault(require("../StatProperty"));
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
/**
 * @en WeaponRefinement
 */
class WeaponRefinement {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        this._data = data;
        this.enka = enka;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.level = json.getAsNumberWithDefault(0, "level") + 1;
        this.name = new TextAssets_1.default(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.default(json.getAsNumber("descTextMapHash"), enka);
        this.addProps = json.get("addProps").filterArray((_, p) => p.has("propType") && p.has("value")).map(([, p]) => new StatProperty_1.default(p.getAsString("propType"), p.getAsNumber("value"), enka));
        this.paramList = json.get("paramList").mapArray((_, p) => p.getAsNumber());
    }
    /**
     * @param id
     * @param level refinement rank (1-5)
     * @param enka
     */
    static getById(id, level, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("EquipAffixExcelConfigData").findArray((_, p) => p.getAsNumber("id") === id && p.getAsNumberWithDefault(0, "level") === level - 1)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("WeaponRefinement", id);
        return new WeaponRefinement(json.getAsJsonObject(), enka);
    }
}
exports.default = WeaponRefinement;
