// enka-network-api/dist/models/weapon/WeaponData.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
const ImageAssets_1 = __importDefault(require("../assets/ImageAssets"));
const TextAssets_1 = __importDefault(require("../assets/TextAssets"));
const WeaponRefinement_1 = __importDefault(require("./WeaponRefinement"));
const StatProperty_1 = __importDefault(require("../StatProperty"));
const WeaponAscension_1 = __importDefault(require("./WeaponAscension"));
const ts_utils_1 = require("../../utils/ts_utils");
/**
 * @en WeaponData
 */
class WeaponData {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        var _a;
        this._data = data;
        this.enka = enka;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.name = new TextAssets_1.default(json.getAsNumber("nameTextMapHash"), enka);
        this.description = new TextAssets_1.default(json.getAsNumber("descTextMapHash"), enka);
        this._nameId = json.getAsString("icon").slice("UI_EquipIcon_".length);
        this.icon = new ImageAssets_1.default(json.getAsString("icon"), enka);
        this.awakenIcon = new ImageAssets_1.default(json.getAsString("awakenIcon"), enka);
        this.stars = json.getAsNumber("rankLevel");
        this.weaponType = json.getAsString("weaponType");
        const weaponTypeData = (_a = enka.cachedAssetsManager.getGenshinCacheData("ManualTextMapConfigData").findArray((_, p) => p.getAsString("textMapId") === this.weaponType)) === null || _a === void 0 ? void 0 : _a[1];
        if (!weaponTypeData)
            throw new AssetsNotFoundError_1.default("Weapon Type", this.weaponType);
        this._weaponTypeData = weaponTypeData.getAsJsonObject();
        this.weaponTypeName = new TextAssets_1.default(weaponTypeData.getAsNumber("textMapContentTextMapHash"), enka);
        this.refinements = (() => {
            const refinementId = json.getAsNumber("skillAffix", 0);
            if (refinementId === 0)
                return [];
            const refinementsJson = enka.cachedAssetsManager.getGenshinCacheData("EquipAffixExcelConfigData").filterArray((_, p) => p.getAsNumber("id") === refinementId).sort(([, a], [, b]) => a.getAsNumberWithDefault(0, "level") - b.getAsNumberWithDefault(0, "level"));
            return refinementsJson.map(([, r]) => new WeaponRefinement_1.default(r.getAsJsonObject(), enka));
        })();
    }
    /**
     * @param ascension ascension level 0-6 for 3-5 stars, and 0-4 for 1-2 stars.
     */
    getAscensionData(ascension) {
        return WeaponAscension_1.default.getById(new config_file_js_1.JsonReader(this._data).getAsNumber("weaponPromoteId"), ascension, this.enka);
    }
    /**
     * @param ascension ascension level 0-6 for 3-5 stars, and 0-4 for 1-2 stars.
     * @param level weapon level 1-90 for 3-5 stars, and 1-70 for 1-2 stars.
     */
    getStats(ascension, level) {
        const maxAscension = this.stars < 3 ? 4 : 6;
        const maxLevel = this.stars < 3 ? 70 : 90;
        if (ascension < 0 || maxAscension < ascension)
            throw new Error(`Ascension levels must be between 0 and ${maxAscension} for ${this.stars} stars.`);
        if (level < 1 || maxLevel < level)
            throw new Error(`Weapon levels must be between 1 and ${maxLevel} for ${this.stars} stars.`);
        const curve = this.enka.cachedAssetsManager.getGenshinCacheData("WeaponCurveExcelConfigData").get(level - 1, "curveInfos");
        const ascensionData = this.getAscensionData(ascension);
        const weaponJson = new config_file_js_1.JsonReader(this._data);
        const weaponProps = weaponJson.get("weaponProp");
        const statPropertiesWithBaseValues = weaponProps.mapArray((_, prop) => {
            var _a, _b, _c;
            const fightProp = prop.getAsStringWithDefault(null, "propType");
            const baseValue = prop.getAsNumberWithDefault(null, "initValue");
            if (!fightProp || !baseValue)
                return null;
            const curveType = prop.getAsString("type");
            const targetCurve = (_a = curve.findArray((__, c) => c.getAsString("type") === curveType)) === null || _a === void 0 ? void 0 : _a[1];
            const ascensionValue = (_c = (_b = ascensionData.addProps.find(p => p.fightProp === fightProp)) === null || _b === void 0 ? void 0 : _b.value) !== null && _c !== void 0 ? _c : 0;
            const value = baseValue * (targetCurve ? targetCurve.getAsNumber("value") : 1) + ascensionValue;
            return new StatProperty_1.default(fightProp, value, this.enka);
        }).filter(ts_utils_1.nonNullable);
        const usedFightProps = statPropertiesWithBaseValues.map(s => s.fightProp);
        const statPropertiesWithoutBaseValues = ascensionData.addProps.filter(p => !usedFightProps.includes(p.fightProp));
        return [...statPropertiesWithBaseValues, ...statPropertiesWithoutBaseValues];
    }
    /**
     * @param id
     * @param enka
     */
    static getById(id, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("WeaponExcelConfigData").findArray((_, p) => p.getAsNumber("id") === id)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("Weapon", id);
        return new WeaponData(json.getAsJsonObject(), enka);
    }
}
exports.default = WeaponData;
