// enka-network-api/dist/models/weapon/WeaponAscension.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const UpgradeCost_1 = __importDefault(require("../material/UpgradeCost"));
const StatProperty_1 = __importDefault(require("../StatProperty"));
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
/**
 * @en WeaponAscension
 */
class WeaponAscension {
    /**
     * @param data
     * @param enka
     */
    constructor(data, enka) {
        this._data = data;
        this.enka = enka;
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("weaponPromoteId");
        this.ascension = json.getAsNumberWithDefault(0, "promoteLevel");
        this.unlockMaxLevel = json.getAsNumber("unlockMaxLevel");
        this.requiredAdventureRank = json.getAsNumberWithDefault(0, "requiredPlayerLevel");
        this.cost = new UpgradeCost_1.default(json.getAsNumberWithDefault(0, "coinCost"), json.has("costItems") ? json.get("costItems").mapArray((_, p) => p.getAsJsonObject()) : [], enka);
        this.addProps = json.get("addProps").filterArray((_, p) => p.has("propType") && p.has("value")).map(([, p]) => new StatProperty_1.default(p.getAsString("propType"), p.getAsNumber("value"), enka));
    }
    /**
     * @param id weaponPromoteId
     * @param ascension promoteLevel
     */
    static getById(id, ascension, enka) {
        var _a;
        const json = (_a = enka.cachedAssetsManager.getGenshinCacheData("WeaponPromoteExcelConfigData").findArray((_, p) => p.getAsNumber("weaponPromoteId") === id && p.getAsNumberWithDefault(0, "promoteLevel") === ascension)) === null || _a === void 0 ? void 0 : _a[1];
        if (!json)
            throw new AssetsNotFoundError_1.default("WeaponAscension", `${id}-${ascension}`);
        return new WeaponAscension(json.getAsJsonObject(), enka);
    }
}
exports.default = WeaponAscension;
