// enka-network-api/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultEnkaClientOptions = exports.isReleased = exports.hasEnergySkill = exports.getNameIdByCharacterId = exports.getCharactersById = exports.validateCache = exports.fetchJSON = exports.StatProperty = exports.CharacterProfilePicture = exports.ProfilePicture = exports.GenshinUser = exports.Element = exports.DetailedGenshinUser = exports.WeaponRefinement = exports.WeaponData = exports.WeaponAscension = exports.Weapon = exports.UpgradeCost = exports.NameCard = exports.Material = exports.GenshinCharacterBuild = exports.Costume = exports.Constellation = exports.CharacterVoiceData = exports.CharacterStats = exports.CharacterDetails = exports.CharacterData = exports.CharacterAscension = exports.Character = exports.UpgradableSkill = exports.UniqueSkill = exports.SkillLevel = exports.Skill = exports.PassiveTalent = exports.NormalAttack = exports.ElementalSkill = exports.ElementalBurst = exports.TextAssets = exports.SkillAttributeAssets = exports.ImageAssets = exports.DynamicTextAssets = exports.ArtifactSplitSubstat = exports.ArtifactSetBonus = exports.ArtifactSet = exports.ArtifactData = exports.Artifact = exports.AssetsNotFoundError = exports.ObjectKeysManager = exports.CachedAssetsManager = exports.EnkaClient = void 0;
exports.GOODUtils = exports.GOOD = exports.percent = exports.fightProps = exports.artifactRarityRangeMap = void 0;
// Avoid circular dependency causes errors.
const CachedAssetsManager_1 = __importDefault(require("./client/CachedAssetsManager"));
exports.CachedAssetsManager = CachedAssetsManager_1.default;
const EnkaClient_1 = __importDefault(require("./client/EnkaClient"));
exports.EnkaClient = EnkaClient_1.default;
const ObjectKeysManager_1 = __importDefault(require("./client/ObjectKeysManager"));
exports.ObjectKeysManager = ObjectKeysManager_1.default;
const AssetsNotFoundError_1 = __importDefault(require("./errors/AssetsNotFoundError"));
exports.AssetsNotFoundError = AssetsNotFoundError_1.default;
const Artifact_1 = __importDefault(require("./models/artifact/Artifact"));
exports.Artifact = Artifact_1.default;
const ArtifactData_1 = __importDefault(require("./models/artifact/ArtifactData"));
exports.ArtifactData = ArtifactData_1.default;
const ArtifactSet_1 = __importDefault(require("./models/artifact/ArtifactSet"));
exports.ArtifactSet = ArtifactSet_1.default;
const ArtifactSetBonus_1 = __importDefault(require("./models/artifact/ArtifactSetBonus"));
exports.ArtifactSetBonus = ArtifactSetBonus_1.default;
const ArtifactSplitSubstat_1 = __importDefault(require("./models/artifact/ArtifactSplitSubstat"));
exports.ArtifactSplitSubstat = ArtifactSplitSubstat_1.default;
const DynamicTextAssets_1 = __importDefault(require("./models/assets/DynamicTextAssets"));
exports.DynamicTextAssets = DynamicTextAssets_1.default;
const ImageAssets_1 = __importDefault(require("./models/assets/ImageAssets"));
exports.ImageAssets = ImageAssets_1.default;
const SkillAttributeAssets_1 = __importDefault(require("./models/assets/SkillAttributeAssets"));
exports.SkillAttributeAssets = SkillAttributeAssets_1.default;
const TextAssets_1 = __importDefault(require("./models/assets/TextAssets"));
exports.TextAssets = TextAssets_1.default;
const ElementalBurst_1 = __importDefault(require("./models/character/talents/ElementalBurst"));
exports.ElementalBurst = ElementalBurst_1.default;
const ElementalSkill_1 = __importDefault(require("./models/character/talents/ElementalSkill"));
exports.ElementalSkill = ElementalSkill_1.default;
const NormalAttack_1 = __importDefault(require("./models/character/talents/NormalAttack"));
exports.NormalAttack = NormalAttack_1.default;
const PassiveTalent_1 = __importDefault(require("./models/character/talents/PassiveTalent"));
exports.PassiveTalent = PassiveTalent_1.default;
const Skill_1 = __importDefault(require("./models/character/talents/Skill"));
exports.Skill = Skill_1.default;
const SkillLevel_1 = __importDefault(require("./models/character/talents/SkillLevel"));
exports.SkillLevel = SkillLevel_1.default;
const UniqueSkill_1 = __importDefault(require("./models/character/talents/UniqueSkill"));
exports.UniqueSkill = UniqueSkill_1.default;
const UpgradableSkill_1 = __importDefault(require("./models/character/talents/UpgradableSkill"));
exports.UpgradableSkill = UpgradableSkill_1.default;
const Character_1 = __importDefault(require("./models/character/Character"));
exports.Character = Character_1.default;
const CharacterAscension_1 = __importDefault(require("./models/character/CharacterAscension"));
exports.CharacterAscension = CharacterAscension_1.default;
const CharacterData_1 = __importDefault(require("./models/character/CharacterData"));
exports.CharacterData = CharacterData_1.default;
const CharacterDetails_1 = __importDefault(require("./models/character/CharacterDetails"));
exports.CharacterDetails = CharacterDetails_1.default;
const CharacterStats_1 = __importDefault(require("./models/character/CharacterStats"));
exports.CharacterStats = CharacterStats_1.default;
const CharacterVoiceData_1 = __importDefault(require("./models/character/CharacterVoiceData"));
exports.CharacterVoiceData = CharacterVoiceData_1.default;
const Constellation_1 = __importDefault(require("./models/character/Constellation"));
exports.Constellation = Constellation_1.default;
const Costume_1 = __importDefault(require("./models/character/Costume"));
exports.Costume = Costume_1.default;
const GenshinCharacterBuild_1 = __importDefault(require("./models/enka/GenshinCharacterBuild"));
exports.GenshinCharacterBuild = GenshinCharacterBuild_1.default;
const Material_1 = __importStar(require("./models/material/Material"));
exports.Material = Material_1.default;
Object.defineProperty(exports, "NameCard", { enumerable: true, get: function () { return Material_1.NameCard; } });
const UpgradeCost_1 = __importDefault(require("./models/material/UpgradeCost"));
exports.UpgradeCost = UpgradeCost_1.default;
const Weapon_1 = __importDefault(require("./models/weapon/Weapon"));
exports.Weapon = Weapon_1.default;
const WeaponAscension_1 = __importDefault(require("./models/weapon/WeaponAscension"));
exports.WeaponAscension = WeaponAscension_1.default;
const WeaponData_1 = __importDefault(require("./models/weapon/WeaponData"));
exports.WeaponData = WeaponData_1.default;
const WeaponRefinement_1 = __importDefault(require("./models/weapon/WeaponRefinement"));
exports.WeaponRefinement = WeaponRefinement_1.default;
const DetailedGenshinUser_1 = __importDefault(require("./models/DetailedGenshinUser"));
exports.DetailedGenshinUser = DetailedGenshinUser_1.default;
const Element_1 = __importDefault(require("./models/Element"));
exports.Element = Element_1.default;
const GenshinUser_1 = __importDefault(require("./models/GenshinUser"));
exports.GenshinUser = GenshinUser_1.default;
const ProfilePicture_1 = __importStar(require("./models/ProfilePicture"));
exports.ProfilePicture = ProfilePicture_1.default;
Object.defineProperty(exports, "CharacterProfilePicture", { enumerable: true, get: function () { return ProfilePicture_1.CharacterProfilePicture; } });
const StatProperty_1 = __importDefault(require("./models/StatProperty"));
exports.StatProperty = StatProperty_1.default;
// functions
var axios_utils_1 = require("./utils/axios_utils");
Object.defineProperty(exports, "fetchJSON", { enumerable: true, get: function () { return axios_utils_1.fetchJSON; } });
var cache_utils_1 = require("./utils/cache_utils");
Object.defineProperty(exports, "validateCache", { enumerable: true, get: function () { return cache_utils_1.validateCache; } });
var character_utils_1 = require("./utils/character_utils");
Object.defineProperty(exports, "getCharactersById", { enumerable: true, get: function () { return character_utils_1.getCharactersById; } });
Object.defineProperty(exports, "getNameIdByCharacterId", { enumerable: true, get: function () { return character_utils_1.getNameIdByCharacterId; } });
Object.defineProperty(exports, "hasEnergySkill", { enumerable: true, get: function () { return character_utils_1.hasEnergySkill; } });
Object.defineProperty(exports, "isReleased", { enumerable: true, get: function () { return character_utils_1.isReleased; } });
// constants
var EnkaClient_2 = require("./client/EnkaClient");
Object.defineProperty(exports, "defaultEnkaClientOptions", { enumerable: true, get: function () { return EnkaClient_2.defaultEnkaClientOptions; } });
var constants_1 = require("./utils/constants");
Object.defineProperty(exports, "artifactRarityRangeMap", { enumerable: true, get: function () { return constants_1.artifactRarityRangeMap; } });
Object.defineProperty(exports, "fightProps", { enumerable: true, get: function () { return constants_1.fightProps; } });
Object.defineProperty(exports, "percent", { enumerable: true, get: function () { return constants_1.percent; } });
// GOOD structures
exports.GOOD = __importStar(require("./models/good/GOOD"));
exports.GOODUtils = __importStar(require("./models/good/IGOODResolvable"));
